### DISCUSSION TITLE
How to make Next not check external backend api when building?

### DISCUSSION THREAD

**Discussion by eNiiju** [vercel/next.js#72395] [Help]
### Summary

I am using Hono inside Deno 2 in my backend with the RPC client.

I import the routes type from the backend app to my Nextjs (Node.js) app (in a single repository):

```ts
import { hc } from 'hono/client';
import type { AppRoutes } from '../../../api-hono/src/app';

export const apiClient = hc<AppRoutes>(process.env.API_PUBLIC_URL!);
```

For the types to be working (instead of having "any"), I had to add this to my nextjs tsconfig.json:
```json
{
  "compilerOptions": {
    "paths": {
      "@/*": ["./src/*"],
      /* ADDED FOR HONO API TYPES (hono is a dev dependency of my nextjs app */
      "hono": ["./node_modules/hono"],
    }
  },
}
```

This works in dev mode, but when building it seems to check the code from my hono app :
```
../api-hono/src/app.ts:2:22
Type error: Cannot find module 'hono/cors' or its corresponding type declarations.

  1 | import { Hono } from "hono";
> 2 | import { cors } from "hono/cors";
    |                      ^
  3 | import { logger } from "hono/logger";
  4 | import { config } from "./config.ts";
  5 | import healthRoute from "./modules/health/healthRoute.ts";
   Linting and checking validity of types  .. ELIFECYCLE  Command failed with exit code 1.
```

Is there a way for Next to not do that ? I just want the type from my backend, for development.

Note that this doesn't happen if I use Hono with Node.js, as I don't need to add anything to the tsconfig.json.

I linked to a minimal reproduction repo.

### Example

https://github.com/eNiiju/debug-nextjs-hono

**Comment by jonas-is-coding** [comment]
The problem you describe occurs because Next.js tries to check all TypeScript and JavaScript files by default during the build process, including those intended for development within the monorepos (in this case, your hono app). So when you try to import the type from the Hono backend app, Next.js checks the code and encounters the error because it tries to find and load the backend modules (like hono/cors) as well.
## Solution
To prevent Next.js from checking the Hono app code during the build process, you can exclude the folder containing the Hono app from Next.js in your tsconfig.json. Simply add the exclude option:
```json
{
  "compilerOptions": {
    "paths": {
      "@/*": ["./src/*"],
      "hono": ["./node_modules/hono"]
    }
  },
  "exclude": [
    "node_modules",
    "api-hono"  // Exclude Hono code from the build process
  ]
}
```
This should ensure that only the types are used without checking the hono/cors code during the build. This should solve the problem.

**Reply by eNiiju** [reply]
Hello, thank you for your answer. Unfortunately, this doesn't seem to change anything, I get the same error when building.

**Reply by jonas-is-coding** [reply]
Okay, after what felt like 100 solution approaches, I have now found a solution, but have run into further errors as a result. 
By adding a **global.d.ts** in the src folder the problem with the hono/cors can be solved.
```javascript
declare module ‘hono/cors’ {
    import { MiddlewareHandler } from ‘hono/cors’;
    export const cors: MiddlewareHandler;
  }
```
I also had to add ```‘allowImportingTsExtensions’: true,``` in the **tsconfig.json** of the nextjs app.
However, since I cloned the repo, Deno is always undefined in main.ts.
```bash
../api/src/main.ts:16:1
Type error: Cannot find name 'Deno'.

  14 |   .route('/health', healthRoute);
  15 |
> 16 | Deno.serve({ port }, app.fetch);
     | ^
  17 |
  18 | export type AppRoutes = typeof routes;
```

**Reply by eNiiju** [reply]
I tried a lot of things too, it's quite frustrating. 

The only thing that worked for building without errors is using:

```ts
const nextConfig = {
    typescript: {
        ignoreBuildErrors: true,
    }
}
```

This is far from ideal though.

**Comment by icyJoseph** [comment]
Don't trust LLM's 😉 

This doesn't get a build running, but it moves forward, onto the next error.

```json
    "paths": {
      "@/*": ["./src/*"],
      /* ADDED FOR HONO API TYPES */
      "hono": ["./node_modules/hono"],
      "hono/cors": ["./node_modules/hono/dist/types/middleware/cors"]
    }
```

I think the issue here is ESM vs CJS and what not... fun stuff


Now, you need to handle the `.ts` stuff...

```json
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    ".next/types/**/*.ts",
    "global.d.ts"
  ],
```

And in `global.d.ts`:

```ts
export declare global {
  export const Deno: any;
}
```

That should get your build rolling ~,but perhaps not properly type-checked? ~

And it is properly type checked at build time. I tried to do a `console.log(apiClient.foo)` within a page, and during build I got:

```console
Property 'foo' does not exist on type '{ index: ClientRequest<{ $get: { input: {}; output: "Hello Hono!"; outputFormat: "text"; status: StatusCode; }; }>; } & { health: ClientRequest<{ $get: { input: {}; output: { status: string; timestamp: string; }; outputFormat: "json"; status: StatusCode; }; }>; }'.
```

**Reply by icyJoseph** [reply]
Alright, edited my answer with the full solution (I think)

**Reply by eNiiju** [reply]
Thank you for your answer, I arrived at the same point too.
Except that instead of declaring Deno as any, I used :
```bash
deno types > lib.deno.d.ts
```
and included that in my `tsconfig.json`.

I only have one more error now, and one that I overlooked when making the example debug project, is that I have JSR imports in my Deno app. And I have no idea how typescript could resolve this from Node./Next.js. This isn't really related to Next.js though.

**Reply by icyJoseph** [reply]
Perhaps, https://www.totaltypescript.com/books/total-typescript-essentials/modules-scripts-and-declaration-files#declare-module ?

**Reply by icyJoseph** [reply]
Also, please don't forget to mark as answered, closed if you are satisfied :)
